### PR TITLE
Create instance creator operation

### PR DESCRIPTION
**Pull request by jejung**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -373,6 +374,13 @@ Returns a function that returns method result for a given object on predefined a
 $userIds = map(methodCaller('getId'), $users);
 ```
 
+##### instanceCreator($class)
+Returns a function that returns a new instance of a predefined class for given parameters
```
Comment (daltones) [inline README.md]: How about `Returns a function that returns a new instance of a predefined class, passing its parameters to the constructor.`? I guess it's more precise.

**Review by daltones — CHANGES_REQUESTED** [review]

File: nspl/op.php [hunk 2]
```diff
@@ -168,3 +168,18 @@ function methodCaller($method, array $args = array())
         return call_user_func_array(array($object, $method), $args);
     };
 }
+
+/**
+ * Returns a function that returns a new instance of a predefined class for given parameters
```
Comment (daltones) [inline nspl/op.php]: Same here.

**Review by daltones — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (jejung) [reply README.md]: Sure, I will fix it

File: nspl/op.php [hunk 3]
```diff
@@ -168,3 +168,18 @@ function methodCaller($method, array $args = array())
...
+ */
+function instanceCreator($class)
+{
+    args\expects(args\string, $class);
+    $reflection_class = new \ReflectionClass($class);
+
+    return function() use ($class, $reflection_class) {
+        return call_user_func_array(array($reflection_class, 'newInstance'), func_get_args());
+    };
+}
```
Comment (ihor) [inline nspl/op.php]: NSPL is focused on speed (which allows us to use this tool in our daily to day projects). There are two optimizations which could be made:
1. Starting PHP 5.6 we have the splat operator ```...```. Using we can get rid of reflection which is relatively slow and have the following code ```return new $class(...func_get_args());```. In my own tests, it makes the speed about two times faster.
2. Again the splat operator is faster than a function call so we can get rid of ```func_get_args()``` too. According to my tests, it makes the code about 2.5 times faster than the original version.

So the final code can be something like that:
```php
function instanceCreator($class)
{
    args\expects(args\string, $class);

    return function(...$args) use ($class) {
        return new $class(...$args);
    };
}
```

We can use ```PHP_VERSION``` constant and ```version_compare``` function to provide a PHP 5.3 solution too.

File: nspl/op.php [hunk 4]
```diff
@@ -168,3 +168,18 @@ function methodCaller($method, array $args = array())
...
+
+/**
+ * Returns a function that returns a new instance of a predefined class, passing its parameters to the constructor
+ * @param string $class Class name
+ * @return callable
+ */
+function instanceCreator($class)
+{
+    args\expects(args\string, $class);
+    $reflection_class = new \ReflectionClass($class);
```
Comment (ihor) [inline nspl/op.php]: Please use camel-cased names ```$reflectionClass```.

**Comment by ihor** [comment]
I like the idea. Please have a look at my comments and make changes.

One thing I want to ask is to change the branch from ```master``` to ```1.2```. This way I'll be able to release this change in ```1.2.1```. Current master version contains lazy evaluations which will be released in ```1.3```.

Also, I would appreciate if you squashed the commits.

Thanks,
Ihor

**Comment by jejung** [comment]
Thanks for your comments @daltones and @ihor, fixed all of them! Awesome to hear that you can apply this to `1.2.1`

**Comment by ihor** [comment]
Merged and released in ```1.2.1```. Thanks for the pull request!

**Pull request merged**